File: leetcode/12-integer-to-roman.rb
LST = [[      1...4,    -1,  'I'],
       [      4...5,    -4, 'IV'],
       [      5...9,    -5,  'V'],
       [     9...10,    -9, 'IX'],
       [    10...40,   -10,  'X'],
       [    40...50,   -40, 'XL'],
       [    50...90,   -50,  'L'],
       [   90...100,   -90, 'XC'],
       [  100...400,  -100,  'C'],
       [  400...500,  -400, 'CD'],
       [  500...900,  -500,  'D'],
       [ 900...1000,  -900, 'CM'],
       [1000...4000, -1000,  'M']]

# @param {Integer} num
# @return {String}
def int_to_roman(num)
  res = ''
  return res if num <= 0
  LST.each do |x|
    if x[0].include? num
      res = x[2] + int_to_roman(num + x[1])
      break
    end
  end
  res
end

File: leetcode/12-integer-to-roman.cpp
/*
Given an integer, convert it to a roman numeral.

Input is guaranteed to be within the range from 1 to 3999.
*/

class Solution {
public:
    string intToRoman(int num) {
        if (num <= 0) {
            return "";
        }
        string result;
        if (num >= 900) {
            return string(num >= 1000 ? "" : "C")
                + "M"
                + intToRoman(num >= 1000 ? num-1000 : num+100-1000);
        }
        if (num >= 400) {
            return string(num >= 500 ? "" : "C")
                + "D"
                + intToRoman(num >= 500 ? num-500 : num+100-500);
        }
        if (num >= 90) {
            return string(num >= 100 ? "" : "X")
                + "C"
                + intToRoman(num >= 100 ? num-100 : num+10-100);
        }
        if (num >= 40) {
            return string(num >= 50 ? "" : "X")
                + "L"
                + intToRoman(num >= 50 ? num-50 : num+10-50);
        }
        if (num >= 9) {
            return string(num >= 10 ? "" : "I")
                + "X"
                + intToRoman(num >= 10 ? num-10 : num+1-10);
        }
        if (num >= 4) {
            return string(num >= 5 ? "" : "I")
                + "V"
                + intToRoman(num >= 5 ? num-5 : num+1-5);
        }
        return "I" + intToRoman(num-1);
    }
};
